### PR TITLE
ENH: .read_pickle(...) from zip containing hidden OS X/macOS metadata files/folders

### LINKED ISSUE #37098: ENH: Support for loading pickles from OS X/macOS zip files containing extraneous "__MACOSX" folder and ".DS_STORE" file
#### Is your feature request related to a problem?

I am trying to load some zipped data from my collaborator into pandas, but the zip was created under a version of OS X that adds an extraneous `__MACOSX` folder in the created zipfile, which causes pandas to error when loading the file. 

The zipfile itself can obviously be easily fixed with e.g. `zip -d filename.zip __MACOSX/\*`, but it may cause a headache for a less experienced user. There is a similary problem involving the hidden file `.DS_STORE` on macOS (see [Wikipedia](https://en.wikipedia.org/wiki/.DS_Store) for more on this), which stores metadata on the user's icon preferences for the contents of the zip...

#### Describe the solution you'd like

I have made the following one line change locally:

https://github.com/pandas-dev/pandas/blob/9cb372376fc78aa66e2559de919592007b74cfaa/pandas/io/common.py#L567

becomes

```python
zip_names = [_ for _ in zf.namelist if not (_.startswith("__MACOSX/") or _.startswith(".DS_STORE"))]
```

Unfortunately this approach will not work for `read_csv(...)` as the compression is handled in the C code.

I can make a PR following the submission of this issue.

#### API breaking implications

Not that I can foresee, the only code path that this effects is changing from a `ValueError` to a correctly loaded the zip file.

#### Describe alternatives you've considered

As mentioned, this could be fixed directly in code before invoking pandas, but I think my suggestion is more convenient and doesn't not require any temporary files or extra disk space.

#### Additional context

Additional context: a lot of scientific data on e.g. figshare suffers from this issue. There may be other hidden files added by other operating systems that could be treated in a similar way.

### PR DESCRIPTION
**Pull request by ml-evs**
- [x] closes #37098 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This PR allows for `.zip` files created by OS X and macOS that contain `__MACOSX` and `.DS_STORE` metadata folders to be loaded by `read_pickle/read_csv/read_table/read_json` without error. 

~It does **not** work with `pd.read_csv(...)` as in that case the compression is handled in the C code. If this enhancement is deemed desirable then I'm willing to have a go at writing it (and the test is already written).~

(No longer true as of https://github.com/pandas-dev/pandas/pull/36997)

Other similar folders may exist from other operating systems, in which case the list could be pulled out as a constant which could be used in the tests and in the module itself.

### CODE REVIEW

**Comment by twoertwein** [comment]
I cannot comment on whether this PR is 'desirable' :)

I think it it should also  work for `pd.read_csv(..., engine="python")`, or? And if I manage to get https://github.com/pandas-dev/pandas/pull/36997 working, it will also work with the c engine.

File: doc/source/whatsnew/v1.1.4.rst [hunk 1]
```diff
@@ -35,7 +35,7 @@ Bug fixes
 
 Other
 ~~~~~
--
+- Added ability to load pickles from ``.zip`` files created by OS X and macOS containing ``__MACOSX/`` or ``.DS_STORE`` hidden folders/files (:issue:`37098`).
```
Comment (twoertwein) [inline doc/source/whatsnew/v1.1.4.rst:38]: Probably should go in 1.2.0

**Comment by ml-evs** [comment]
> I cannot comment on whether this PR is 'desirable' :)

I wouldn't want to argue particularly hard for it myself! I guess it depends how prevalent this "problem" is, but I won't be too upset if this doesn't make it in :upside_down_face: 

> I think it it should also work for `pd.read_csv(..., engine="python")`, or? And if I manage to get #36997 working, it will also work with the c engine.

It does indeed work with the Python engine, the tests in `tests/io/parsers/test_compression.py` are run on every engine, I'm happy to push the failing tests if that would useful. Can also rebase onto your PR and check if that works too. 

I'll hold off until there's been a discussion about whether this is desirable (either here or in the issue #37098).

**Comment by twoertwein** [comment]
> It does indeed work with the Python engine, the tests in `tests/io/parsers/test_compression.py` are run on every engine, I'm happy to push the failing tests if that would useful. Can also rebase onto your PR and check if that works too.
> 
> I'll hold off until there's been a discussion about whether this is desirable (either here or in the issue #37098).

I wouldn't wait for my PR (I'm not sure how long it will take me to debug two failing tests on windows). I will ping you when/if the PR gets merged. Including test for the python engine would be good.

File: pandas/tests/io/test_pickle.py [hunk 2]
```diff
@@ -376,6 +376,40 @@ def test_read_infer(self, ext, get_random_path):
 
             tm.assert_frame_equal(df, df2)
 
+    @pytest.mark.parametrize("cruft", ["__MACOSX/", ".DS_STORE"])
+    def test_load_zip_with_hidden_folders(self, cruft, get_random_path):
+        """Test loading .zip files that have extraneous metadata in hidden folders. """
+        base = get_random_path
+        path1 = base + ".raw"
+        path2 = base + ".zip"
+        dummy = base + ".dummy"
```
Comment (twoertwein) [inline pandas/tests/io/test_pickle.py:385]: pandas prefers f-strings: f"{base}.raw"

Comment (ml-evs) [reply pandas/tests/io/test_pickle.py]: Done, I didn't go as far as replacing it for all other tests in this file

File: pandas/tests/io/test_pickle.py [hunk 3]
```diff
@@ -376,6 +376,40 @@ def test_read_infer(self, ext, get_random_path):
 
             tm.assert_frame_equal(df, df2)
 
+    @pytest.mark.parametrize("cruft", ["__MACOSX/", ".DS_STORE"])
+    def test_load_zip_with_hidden_folders(self, cruft, get_random_path):
+        """Test loading .zip files that have extraneous metadata in hidden folders. """
```
Comment (jbrockmendel) [inline pandas/tests/io/test_pickle.py:381]: comment instead of docstring

File: pandas/tests/io/test_pickle.py [hunk 4]
```diff
@@ -376,6 +376,40 @@ def test_read_infer(self, ext, get_random_path):
...
+        base = get_random_path
+        path1 = f"{base}.raw"
+        path2 = f"{base}.zip"
+        dummy = f"{base}.dummy"
+        compression = "zip"
+
+        with tm.ensure_clean(path1) as p1, tm.ensure_clean(
+            path2
+        ) as p2, tm.ensure_clean(dummy) as d:
```
Comment (jbrockmendel) [inline pandas/tests/io/test_pickle.py:390]: can you use a longer variable name than d

File: pandas/tests/io/test_pickle.py [hunk 5]
```diff
@@ -376,6 +376,40 @@ def test_read_infer(self, ext, get_random_path):
...
+            df = tm.makeDataFrame()
+
+            # write to uncompressed file
```
Comment (WillAyd) [inline pandas/tests/io/test_pickle.py:393]: You can remove all of these comments

**Review by WillAyd — CHANGES_REQUESTED** [review]
Is this only an issue on read? Not write?

**Comment by ml-evs** [comment]
> Is this only an issue on read? Not write?

I think it's only a problem when the zip has been made inside the macOS file manager, rather than say, at the command-line (though I do not have a Mac to test this).

As I said before, maybe this is a non-issue.

File: pandas/tests/io/test_pickle.py (hunk 5 above)
Comment (ml-evs) [reply pandas/tests/io/test_pickle.py]: Really? I'm just following the style of other tests in this file. I've made the comments more succinct but wouldn't want to remove them altogether as this is a bit of a weird edge-case to test.

**Comment by twoertwein** [comment]
@ml-evs #36997 is merged: you PR will also work for `read_csv` (both c engine and python engine).

**Comment by ml-evs** [comment]
> @ml-evs #36997 is merged: you PR will also work for `read_csv` (both c engine and python engine).

This seems to be the case, thanks @twoertwein!

**Comment by ml-evs** [comment]
Happy to close this if there's no interest; will re-ping the initial reviewers just in case.

File: doc/source/whatsnew/v1.2.0.rst [hunk 6]
```diff
@@ -253,6 +253,8 @@ Other enhancements
 - Improve error reporting for :meth:`DataFrame.merge()` when invalid merge column definitions were given (:issue:`16228`)
 - Improve numerical stability for :meth:`Rolling.skew()`, :meth:`Rolling.kurt()`, :meth:`Expanding.skew()` and :meth:`Expanding.kurt()` through implementation of Kahan summation (:issue:`6929`)
 - Improved error reporting for subsetting columns of a :class:`DataFrameGroupBy` with ``axis=1`` (:issue:`37725`)
+- :func:`pandas.load_pickle` can now load from ``.zip`` files created by OS X/macOS that contain ``__MACOSX/`` or ``.DS_STORE`` hidden folders/files (:issue:`37098`).
```
Comment (twoertwein) [inline doc/source/whatsnew/v1.2.0.rst:256]: should this be `read_pickle` and also any other `read_*` that supports zip compression? I think it should also be `:func:`read_pickle`` (no `pandas.`)

Comment (ml-evs) [reply doc/source/whatsnew/v1.2.0.rst:256]: Yep, you're right. I think that means just `read_csv`, `read_json`, `read_pickle` and `read_table`.

```suggestion
- :func:`read_pickle` (and other `read_*` functions that handle compressed inputs) can now load from ``.zip`` files created by OS X/macOS that contain ``__MACOSX/`` or ``.DS_STORE`` hidden folders/files (:issue:`37098`).
```

**Comment by jbrockmendel** [comment]
ive tried, but have been unable to form an opinion

**Comment by twoertwein** [comment]
a more generic approach might be to use `{'method': 'zip', 'file': 'path/within/zip'}` to let the user choose which file in a ZIP file should be used? This is more work for the user but allows to read different files within a ZIP file.

Technically this can already be achieve with 
```py
from zipfile import ZipFile

import pandas as pd

with ZipFile("test.zip", mode="r") as file:
    with file.open("data.pickle", mode="r") as file:
        pd.read_pickle(file)
```

**Pull request closed without merging**